optics_functions/coupling.py

```diff
@@ -0,0 +1,208 @@
+    """ Returns the coupling term.
+
+    .. warning::
+        This function changes sign of the real part of the RDTs compared to
+        [FranchiAnalyticFormulas2017]_ to be consistent with the RDT
+        calculations from [CalagaBetatronCoupling2005]_ .
+
+    Args:
+        df (TfsDataFrame): Twiss Dataframe
+        real (bool): Split complex columns into two real-valued columns.
+
+    Keyword Args:
+        **kwargs: Remaining arguments from :func:`~optics_functions.rdt.rdts`
+                  i.e. ``qx``, ``qy``, ``feeddown``, ``loop_phases``
+                  and ``hamiltionian_terms``
+
+    Returns:
+        New TfsDataFrame with Coupling Columns
+    """
+    df_res = calculate_rdts(df, rdts=COUPLING_RDTS, **kwargs)
+    for rdt in COUPLING_RDTS:
+        df_res.loc[:, rdt].to_numpy().real *= -1  # definition, also: sets value in dataframe
+
+    if real:
+        df_res = split_complex_columns(df_res, COUPLING_RDTS)
+
+    return df_res
+
+
+def coupling_from_cmatrix(df: TfsDataFrame, real=False,
+                          output: Sequence[str] = ("rdts", "gamma", "cmatrix")):
+    """ Calculates C matrix and Coupling and Gamma from it.
+    See [CalagaBetatronCoupling2005]_
+
+    Args:
+        df (TfsDataFrame): Twiss Dataframe
+        real (bool): Split complex columns into two real-valued columns.
+        output (list[str]): Combination of 'rdts', 'gamma' and 'cmatrix'.
+                            Specifies which parameters one wants to output.
+
+    Returns:
+        New TfsDataFrame with columns as specified in 'output'.
+    """
+    LOG.debug("Calculating CMatrix.")
+    df_res = TfsDataFrame(index=df.index)
+    with timeit("CMatrix calculation", print_fun=LOG.debug):
+        n = len(df)
+        gx, r, igy = np.zeros((n, 2, 2)), np.zeros((n, 2, 2)), np.zeros((n, 2, 2))
+
+        # rs form after -J R^T J == inv(R)*det|R| == C
+        r[:, 0, 0] = df["R22"]
+        r[:, 0, 1] = -df["R12"]
+        r[:, 1, 0] = -df["R21"]
+        r[:, 1, 1] = df["R11"]
+
+        r *= 1 / np.sqrt(1 + np.linalg.det(r)[:, None, None])
+
+        # Cbar = Gx * C * Gy^-1
+        sqrtbetax = np.sqrt(df[f"{BETA}{X}"])
+        sqrtbetay = np.sqrt(df[f"{BETA}{Y}"])
+
+        gx[:, 0, 0] = 1 / sqrtbetax
+        gx[:, 1, 0] = df[f"{ALPHA}{X}"] * gx[:, 0, 0]
+        gx[:, 1, 1] = sqrtbetax
+
+        igy[:, 1, 1] = 1 / sqrtbetay
+        igy[:, 1, 0] = -df[f"{ALPHA}{Y}"] * igy[:, 1, 1]
+        igy[:, 0, 0] = sqrtbetay
+
+        c = np.matmul(gx, np.matmul(r, igy))
+        gamma = np.sqrt(1 - np.linalg.det(c))
+
+    if "rdts" in output:
+        denom = 1 / (4 * gamma)
+        df_res.loc[:, "F1001"] = ((c[:, 0, 0] + c[:, 1, 1]) * 1j +
+                                  (c[:, 0, 1] - c[:, 1, 0])) * denom
+        df_res.loc[:, "F1010"] = ((c[:, 0, 0] - c[:, 1, 1]) * 1j +
+                                  (-c[:, 0, 1]) - c[:, 1, 0]) * denom
+        LOG.info(f"Average coupling amplitude |F1001|: {df_res['F1001'].abs().mean():g}")
```

Comment:
maybe this should be debug like the rest of the function's logging

Reply:
nah, this is the result. That's the only thing work informing about. That's why the rest is debug

Reply:
maybe the start of the function could be info as well. 
I like to print info before things that might take a moment, so the user knows why the computer is heating up

Reply:
Very fair, let's make the first one `info` and resolve this conversation.